pkg/liquidity-source/mx-trading/pool_simulator.go

```diff
@@ -0,0 +1,190 @@
+
+		token0, token1 entity.PoolToken
+
+		timestamp int64
+		gas       Gas
+	}
+)
+
+func NewPoolSimulator(entityPool entity.Pool) (*PoolSimulator, error) {
+	var extra PoolExtra
+	if err := json.Unmarshal([]byte(entityPool.Extra), &extra); err != nil {
+		return nil, err
+	}
+
+	return &PoolSimulator{
+		Pool: pool.Pool{
+			Info: pool.PoolInfo{
+				Address:    strings.ToLower(entityPool.Address),
+				ReserveUsd: entityPool.ReserveUsd,
+				Exchange:   entityPool.Exchange,
+				Type:       entityPool.Type,
+				Tokens: lo.Map(entityPool.Tokens,
+					func(item *entity.PoolToken, index int) string { return item.Address }),
+				Reserves: lo.Map(entityPool.Reserves,
+					func(item string, index int) *big.Int { return bignumber.NewBig(item) }),
+			},
+		},
+		ZeroToOnePriceLevels: extra.ZeroToOnePriceLevels,
+		OneToZeroPriceLevels: extra.OneToZeroPriceLevels,
+
+		token0:    *entityPool.Tokens[0],
+		token1:    *entityPool.Tokens[1],
+		timestamp: entityPool.Timestamp,
+		gas:       defaultGas,
+	}, nil
+}
+
+func (p *PoolSimulator) CalcAmountOut(params pool.CalcAmountOutParams) (*pool.CalcAmountOutResult, error) {
+	if params.TokenAmountIn.Token == p.token0.Address {
+		return p.swap(params.TokenAmountIn.Amount, p.token0, p.token1,
+			params.Limit.GetLimit(p.token1.Address), p.ZeroToOnePriceLevels,
+		)
+	} else {
+		return p.swap(params.TokenAmountIn.Amount, p.token1, p.token0,
+			params.Limit.GetLimit(p.token0.Address), p.OneToZeroPriceLevels,
+		)
+	}
+}
+
+func (p *PoolSimulator) swap(
+	amountIn *big.Int,
+	baseToken, quoteToken entity.PoolToken,
+	inventoryLimit *big.Int,
+	priceLevel []PriceLevel,
+) (*pool.CalcAmountOutResult, error) {
+	amountInF, _ := amountIn.Float64()
+	amountInAfterDecimalsF := amountInF / math.Pow10(int(baseToken.Decimals))
+	fillPrice, err := findFillPrice(amountInAfterDecimalsF, priceLevel)
+	if err != nil {
+		return nil, err
+	}
+	amountOutAfterDecimalsF := amountInAfterDecimalsF * fillPrice
+	amountOutF := amountOutAfterDecimalsF * math.Pow10(int(quoteToken.Decimals))
+	amountOut, _ := big.NewFloat(amountOutF).Int(nil)
+
+	if amountOut.Cmp(inventoryLimit) > 0 {
+		return nil, ErrAmountOutIsGreaterThanInventory
+	}
+
+	return &pool.CalcAmountOutResult{
+		TokenAmountOut: &pool.TokenAmount{Token: quoteToken.Address, Amount: amountOut},
+		Fee:            &pool.TokenAmount{Token: baseToken.Address, Amount: bignumber.ZeroBI},
+		Gas:            p.gas.FillOrderArgs,
+		SwapInfo: SwapInfo{
+			BaseToken:       baseToken.Address,
+			BaseTokenAmount: amountIn.String(),
+			QuoteToken:      quoteToken.Address,
+		},
+	}, nil
+}
+
+func (p *PoolSimulator) UpdateBalance(params pool.UpdateBalanceParams) {
+	tokenIn, tokenOut := params.TokenAmountIn.Token, params.TokenAmountOut.Token
+	amountIn, amountOut := params.TokenAmountIn.Amount, params.TokenAmountOut.Amount
+	amountInF, _ := amountIn.Float64()
+
+	var (
+		amountInAfterDecimalsF float64
+		priceLevels            *[]PriceLevel
```

Comment:
any reason not to use
```suggestion
		priceLevels            []PriceLevel
```
?

Reply:
updated at [deb889a](https://github.com/KyberNetwork/kyberswap-dex-lib/pull/605/commits/deb889a6c2b281e66135106ebb72f6cee807b6d6)